arch/risc-v/src/litex/litex_sdio.c

```diff
@@ -903,7 +905,9 @@ static int litex_sendsetup(struct sdio_dev_s *dev,
 
   /* flush CPU d-cache */
 
+#ifndef CONFIG_LITEX_COHERENT_DMA
   up_invalidate_dcache_all();
```

Comment:
I think that the issue may be also hidden here. The recv path definitely needs `invalidate`, but I think that send path either needs `clean` or `flush` and `invalidate` will destroy data that are currently in cache, so will send whatever is in memory and that seems to be wrong.
I suggest to test with `up_invalidate_dcache(buffer, buffer + nbytes);` in `litex_recvsetup()` and with `up_clean_dcache(buffer, buffer + nbytes)` in `litex_sendsetup()`